h2o-core/src/main/java/hex/Model.java

```diff
@@ -53,6 +53,10 @@
   protected ScoringInfo[] scoringInfo;
   public IcedHashMap<Key, String> _toDelete = new IcedHashMap<>();
 
+  public enum Type {
+    PREDICTIVE, DATA_TRANSFORMING
```

Comment:
From now on, models won't be just predictors, but also data transformers. There is no transform() method yet, but this is just a matter of API changes right now. The method `predict` is going to behave as `transform` for now.

At several places, there is a need to know what kind of model we're working with. E.g. to create a name for the destination frame, where for data transformers, it is not desired to have the resulting data frame named with prefix "predictions_".

Reply:
@Pscheidl I would suggest to separate these two types by leveraging class hierarchy. This would allow us to keep code that is specific to a particular model type in a separate class/hierarchy branch. With just `enum` the separation is not strong enough and everything will be in one place/class. Just thinking out loud.  @michalkurka  wdyt?

Reply:
How'd you want to do that, without throwing away everything from current Model class ?
Currently, there is a clear hierarchy. The new one would require creation of new services around a new type of model, including frontend code generation mechanisms and other stuff. 

 The plan is to bring a new general method named `transform` to current models, so that it can be used from now on and delegate it's behavior to former `predict` or `score` where necessary (in all previous predictive models). This way, it won't take forever to implement it.

Btw. you can see not transform method yet, because it's not yet implemented, as this PR is a Work In Progress currently.

Reply:
@Pscheidl Why throwing away? Moving is not throwing. What if we had code generation for one class that is doing everything in our app.... sooner or later we would want to leverage benefits of OOP and introduce new classes.  I'm just suggesting **to consider** this option....because what you are doing is just a half step. The fact that there are a lot of things that needs to be changed is an obstacle of course but it doesn't mean that what I'm suggesting is a wrong thing. Though it might be a wrong thing but it should not be argued against with just time expenses.
Regarding general method that will delegate to either score or transform based on enum's value.... is a sign that you are combining two things that you already have acknowledged to be a distinguishable things. You are increasing complexity of the class Model. OOP approach is to expand hierarchy and make it easier to reason as soon as we have opportunity to separate things. Having `transform` and `score` methods in two separate classes that are calling same parent's method - lets say 'process` or again `transform` -would be better IMHO.
I know that this PR is in progress but you are writing comments for others and I think that discussing them along the way is better than at the end of the work.

Reply:
The transform method will exist algonside predict and score, not replacing it. Nor there will be a router (at the moment). Again, we're talking about WIP here.

Reply:
The complexity of such task is non-trivial. This has various reasons and there are not few.

Most of the current algos somehow depend on current `Model` class. What is there may not be used by all, but every part is used by some of the underlying implementations.  H2O is designed this way. And this affects the whole core and algos module related to models. For example parameters. Some of the parameters are unused by some of the models (obviously GenericModel for example). This way, we might squeeze any transformation machine in.

Creating a new predecessor only for transformations with a common method for transformations/predictions with existing models, one would have to do something like:

```
   Transformation
      <                 >
   <                      >
Model                  DataTransformation
```

Which effectively means creating new frontend generators, new V3 generators and checks (or new piece of Rest API to get rid of V3 objects if required). New frontend validation and display logic (or reuse it and wire it, not trivial in any case). Make it compatible with Flow. And test it all.

This is what I wanted to avoid, as I still believe a Model can still describe a universal transformation and the `train` and `computeImpl` methods actually allow us to do any customizations we want internally. All that needs to be done is to expose new `transform` function, that will be usable both for predicting and for data transformation. On the client. I find this much easier, much less error prone and much more likely to deliver in time. 

Your solution sounds nicer, of course. But it basically means writing and re-writing lots of code, which requires lots of time, testing and eventually bugfixing.

Reply:
@Pscheidl As usual, just by always following the easiest and fastest way eventually you will find yourself in a big unbearable trouble. I'm sure you know about this. All what I wanted is to bring up/discuss another option. I believe that just by bearing in mind multiple options we can end up with a better solution.

Reply:
@Pscheidl I know the discussion is outdated, but FYI  it's hard not to consider that `AggregatorModel`, `SVDModel`, `PCAModel`, `GLRMModel` are already actual transformers with an unfortunate `predict` API.

Reply:
@seb-h2o Let's create other JIRA once this new transform API is merged. I'll take care of it. Great point, thank you.